File: examples/bubble_sort.go
package main

import "fmt"

// bubbleSort sorts an array of integers using bubble sort algorithm
func bubbleSort(arr []int) []int {
	n := len(arr)
	for i := 0; i < n; i++ {
		for j := 0; j < n-i-1; j++ {
			if arr[j] > arr[j+1] {
				// Swap elements
				arr[j], arr[j+1] = arr[j+1], arr[j]
			}
		}
	}
	return arr
}

func main() {
	numbers := []int{64, 34, 25, 12, 22, 11, 90}
	fmt.Println("Original array:", numbers)

	// Create a copy to preserve original
	sorted := make([]int, len(numbers))
	copy(sorted, numbers)
	bubbleSort(sorted)

	fmt.Println("Sorted array:", sorted)
}

File: examples/bubble_sort.rs
/// Bubble sort implementation in Rust
fn bubble_sort(arr: &mut [i32]) {
    let n = arr.len();
    for i in 0..n {
        for j in 0..n - i - 1 {
            if arr[j] > arr[j + 1] {
                // Swap elements
                arr.swap(j, j + 1);
            }
        }
    }
}

fn main() {
    let mut numbers = vec![64, 34, 25, 12, 22, 11, 90];
    println!("Original array: {:?}", numbers);

    bubble_sort(&mut numbers);

    println!("Sorted array: {:?}", numbers);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_bubble_sort() {
        let mut arr = vec![5, 2, 8, 1, 9];
        bubble_sort(&mut arr);
        assert_eq!(arr, vec![1, 2, 5, 8, 9]);
    }
}
